statedb/statedb.go

```diff
@@ -130,49 +123,55 @@ func NewSubTreeConfig(hashFunc arbo.HashFunction, kindID string, maxLevels int,
 	}
 }
 
-// treeConfig returns a unified configuration type for opening a singleton
-// subTree that is identified by `id`.  `id` is the path in the parent tree to
-// the leaf that contains the subTree root.
-func (c *SubTreeConfig) treeConfig(id []byte) *treeConfig {
-	return &treeConfig{
-		parentLeafKey: id,
-		prefix:        c.kindID + string(id),
-		SubTreeConfig: c,
+// HashFunc returns the hashFunc set for this SubTreeConfig
+func (c *TreeNonSingletonConfig) HashFunc() arbo.HashFunction {
+	return c.hashFunc
+}
+
+// WithKey returns a unified subTree configuration type for opening a singleton
+// subTree that is identified by `key`.  `key` is the path in the parent tree
+// to the leaf that contains the subTree root.
+func (c *TreeNonSingletonConfig) WithKey(key []byte) *TreeConfig {
+	return &TreeConfig{
+		parentLeafKey:          key,
+		prefix:                 c.kindID + string(key),
+		TreeNonSingletonConfig: c,
 	}
 }
 
-// SubTreeSingleConfig contains the configuration used for a singleton subTree.
-type SubTreeSingleConfig struct {
-	*SubTreeConfig
+// TreeConfig contains the configuration used for a subTree.
+type TreeConfig struct {
+	parentLeafKey []byte
+	prefix        string
+	*TreeNonSingletonConfig
 }
```

Comment:
TreeConfig is just a pointer, a string, and a slice: roughly the size of six pointers. I'd just use it without a pointer like `TreeConfig` rather than `*TreeConfig`, because unnecessary pointers to small types just adds overhead. Unless you want functions to be able to omit the config via `fn(nil)` rather than `fn(TreeConfig{})`, but I don't think that is the case.

This would be different if TreeConfig implemented an interface - since interfaces in Go always contain pointers, trying to avoid the pointer is futile. But I also don't think that's the case here.

This might apply to other small types in this PR, I'm not sure. Just something for you to think about. If a type is small, and doesn't benefit from being used as a pointer, don't use one. And "small" is bigger than you think these days, thanks to the larger caches and number of registers. 32, 64, or even 128 bytes without a pointer are all reasonable.